8266310: deadlock between System.loadLibrary and JNI FindClass loading another class

voitylov opened the pull request:
Please review this PR which fixes the deadlock in ClassLoader between the two lock objects - a lock object associated with the class being loaded, and the ClassLoader.loadedLibraryNames hash map, locked during the native library load operation.

Problem being fixed:

The initial reproducer demonstrated a deadlock between the JarFile/ZipFile and the hash map. That deadlock exists even when the ZipFile/JarFile lock is removed because there's another lock object in the class loader, associated with the name of the class being loaded. Such objects are stored in ClassLoader.parallelLockMap. The deadlock occurs when JNI_OnLoad() loads exactly the same class, whose signature is being verified in another thread.

Proposed fix:

The proposed patch suggests to get rid of locking loadedLibraryNames hash map and synchronize on each entry name, as it's done with class names in see ClassLoader.getClassLoadingLock(name) method.

The patch introduces nativeLibraryLockMap which holds the lock objects for each library name, and the getNativeLibraryLock() private method is used to lazily initialize the corresponding lock object. nativeLibraryContext was changed to ThreadLocal, so that in any concurrent thread it would have a NativeLibrary object on top of the stack, that's being currently loaded/unloaded in that thread. nativeLibraryLockMap accumulates the names of all native libraries loaded - in line with class loading code, it is not explicitly cleared.

Testing:  jtreg and jck testing with no regressions. A new regression test was developed.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266310](https://bugs.openjdk.java.net/browse/JDK-8266310): deadlock between System.loadLibrary and JNI FindClass loading another class


### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 85005d57246f3449d67228c8bf06689704e4721d
 * [Peter Levart](https://openjdk.java.net/census#plevart) (@plevart - **Reviewer**) ⚠️ Review applies to 85005d57246f3449d67228c8bf06689704e4721d
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to 85005d57246f3449d67228c8bf06689704e4721d
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3976/head:pull/3976` \
`$ git checkout pull/3976`

Update a local copy of the PR: \
`$ git checkout pull/3976` \
`$ git pull https://git.openjdk.java.net/jdk pull/3976/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3976`

View PR using the GUI difftool: \
`$ git pr show -t 3976`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3976.diff">https://git.openjdk.java.net/jdk/pull/3976.diff</a>

</details>

voitylov commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -456,20 +458,43 @@ public void run() {
+    }
+
+    // thread local native libraries stack
+    private static final ThreadLocal<Deque<NativeLibraryImpl>> nativeLibraryThreadContext =
```
I will update the PR to avoid ThreadLocal and have ArrayDeque deallocated once its last element has been popped out.

plevart commented:
Hi Aleksei,
If I understand David Holmes correctly (on the mailing list), there is a concern that using a per-library-name exclusive lock might not be enough to prevent deadlock with class loading lock. You say: "The deadlock occurs when JNI_OnLoad() loads exactly the same class, whose signature is being verified in another thread." ... Could you explain a little bit more about this scenario? As I understand the JNI_OnLoad() is called as part of native library loading and at that time the thread that is loading native library holds a nativeLibraryLock(libraryName). Now if JNI_OnLoad() triggers loading of a class, whose signature is being verified in another thread (do you mean: which is already being loaded in another thread?) then JNI_OnLoad() thread will block on classLoadingLock(className). But that is not a dead-lock yet. The loading of a class in another thread now has to attempt to load the same native library that is being loaded in the JNI_OnLoad() thread.
So before the proposed patch it was enough for the other thread which was loading the class to trigger loading of any native library, while after the proposed patch, the deadlock would occur only if it triggered loading of the same native library. Am I correctly assessing the problem?

voitylov commented:
Peter, that's right. It is still possible to have the same pair of objects being locked with per-library lock, namely when a library being loaded explicitly loads a class that requires the same library to be loaded. I think this type of design flaw cannot be solved by changing locking algorithm. It's clearly not the goal of this patch.

ChrisHegarty commented:
Hi Aleksei, 

As you may know, I looked into a similar issue recently and put together a reproducer [1] (which is probably similar to what you have). The reproducer, run at the time against Oracle 11u, demonstrates the issue on the mainline too, but the deadlock is slightly different.   The reason I mention it here is that the reproducer encounters the issue whether there is an attempt to load the same class or another class ( from the same jar file ).  In fact, the issue is even more general, the problem is with trying to load a class, not already loaded, from a jar further down on the class path ( the class may not even exist, just that it causes the loader to walk the class path up to the jar being verified ).

I filed an issue for this, which may need to be closed as a duplicate depending on the outcome of this PR [2].

[1] https://github.com/ChrisHegarty/deadlock
[2] https://bugs.openjdk.java.net/browse/JDK-8266350

voitylov commented:
Dear colleagues,

The updated PR addresses review comment regarding ThreadLocal as well as David' concern around the lock being held during JNI_OnLoad/JNI_OnUnload calls, and ensures all lock objects are deallocated. Multiple threads are allowed to enter NativeLibrary.load() to prevent any thread from locking while another thread loads a library. Before the update, there could be a class loading lock held by a parallel capable class loader, which can deadlock with the library loading lock. As proposed by David Holmes, the library loading lock was removed because dlopen/LoadLibrary are thread safe and they maintain internal reference counters on libraries. There's still a lock being held while a pair of containers are read/updated. It's not going to deadlock as there's no lock/wait operation performed while that lock is held. Multiple threads may create their own copies of NativeLibrary object and register it for auto unloading. 

Tests for auto unloading were added along with the PR update. There are now 3 jtreg tests:
- one checks for deadlock, similar to the one proposed by Chris Hegarty
- two other tests are for library unload. 

The major side effect of that multiple threads are allowed to enter is that JNI_OnLoad/JNI_OnUnload may be called multiple (but same) number of times from concurrent threads. In particular, the number of calls to JNI_OnLoad must be equal to the number of calls to JNI_OnUnload after the relevant class loader is garbage collected. This may affect the behaviour that relies on specific order or the number of JNI_OnLoad/JNI_OnUnload calls. The current JNI specification does not mandate how many times JNI_OnLoad/JNI_OnUnload are called. Also, we could not locate tests in jck/jtreg/vmTestbase that would rely on the specific order or number of calls to JNI_OnLoad/JNI_OnUnload. 

Thank you Alan Bateman, David Holmes and Chris Hegarty for your valuable input.

JornVernee commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -196,57 +198,67 @@ private NativeLibrary loadLibrary(Class<?> fromClass, String name, boolean isBui
             return lib;
+        } finally {
+            releaseNativeLibraryLock(name);
         }
```
The new locking scheme looks incorrect to me. It now seems possible for 2 different class loaders in 2 different threads to load the same library (which should not be possible).

Thread 1:
  - acquires name lock
  - checks library name is already in `loadedLibraryNames` (it's not)
  - release name lock
  - start loading library
  
Then thread 2 comes in and does the same
  
Then again thread 1 finishes loading and:
 - acquires name lock
 - puts library name in `loadedLibraryNames`
 - puts library name in it's own `libraries`
 - release lock
 
Then thread 2 comes in and does the same again (although adding the name to `loadedLibraryNames` will silently return `false`).

It seems that this scenario is possible, and the library will be loaded by 2 different class loaders, each with their own `lib` object. (If there's a race, there has to be a loser as well, which would need to discard their work when finding out they lost)

You might be able to stress this by checking if `loadedLibraryNames.add(name);` returns `true`, and throwing an exception if not.

I think a possible solution would be to claim the library name up front for a particular loader. Then 2 threads can still race to load the same library for the same class loader, but 2 threads with 2 different class loaders racing to load the library should not be possible.

Actually, you might be able to prevent a race on JNI_OnLoad altogether by claiming the library name for a particular thread upfront as well (e.g. using a `ConcurrentHashMap<String, Thread>`).

voitylov replied:
Thank you, the previous version of the PR suffered from other problems as well. 

The PR is now reverted to a scheme with a lock held on library name, thus such a situation is no longer possible.

JornVernee replied:
Ok, thanks.

FWIW, I think trying to lock on a per-name basis instead of globally is a good idea, and should improve the current situation. It is also safe to do as far as I can see.

voitylov commented:
Sorry, let me get back to this after I talk with the chalkboard.

voitylov commented:
The PR was updated as follows:

- the fix resolves the deadlock problem when different libraries are loaded. 
- the lock on the library name is held just like in the first version of PR to avoid simultaneous JNI_OnLoad calls. 
- the test now checks that JNI_Onload and JNI_OnUnload are executed only once.

Within the current limitations it is not possible to resolve the deadlock when a library being loaded attempts, in JNI_OnLoad,  to load a class that depends on the same library. It is not the intent of this PR to do so.

dholmes-ora commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -456,20 +465,109 @@ public void run() {
+
+    private static void releaseNativeLibraryLock(String libraryName) {
+        CountedLock lock = nativeLibraryLockMap.computeIfPresent(libraryName, (name, currentLock) -> {
+            if (currentLock.getCounter() == 1) {
+                // unlock and release the object if no other threads are queued
+                currentLock.unlock();
```
This isn't atomic so I don't see how it can work as desired. Overall I don't understand what the semantics of this "counted lock" are intended to be.

voitylov replied:
Hi David,

The locking strategy is as follows:

CountedLock is a subclass of ReentrantLock that allows exact counting of threads that intend to acquire the lock object. Each time a thread calls acquireNativeLibraryLock() with a certain name, either a new CountedLock object is allocated and assigned 1 as the counter, or an existing CountedLock is incremented prior to invocation of the lock() method on it. The increment operation to the lock object is performed in the context of execution of compute() method, which is executed atomically. This allows to correctly dispose the CountedLock object when the last thread in the queue leaves releaseNativeLibraryLock(). The entire remapping function passed to computeIfPresent() method is executed atomically. 

Could you be more specific on what is not performed atomically?

dholmes-ora reviewed (commented):
Sorry Aleksei but your counted lock has me a bit mystified. Can you elaborate on the locking strategy please.

Thanks,
David

dholmes-ora commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -456,20 +465,109 @@ public void run() {
+            new ConcurrentHashMap<>();
+
+    private static void acquireNativeLibraryLock(String libraryName) {
+        nativeLibraryLockMap.compute(libraryName, (name, lock) -> {
```
This would be clearer if lock were named currentLock as for releaseNLLock

voitylov replied:
David, lock is now renamed to currentLock

dholmes-ora reviewed (approved):
Hi Aleksei,

Thanks for bearing with me while I got my head round the synchronization protocol - piggy-backing on the CHM logic is clever! That locking part all makes sense to me now and I think it is a good solution to reduce the scope of the deadlock potential.

I'm less clear on the changes to the NativeLibraryContext as I'm not certain what it means today, so making it a per-thread mapping is not something I can really comment on. I'll leave that aspect to core-libs folk. So my approval should not be taken as an approval to push yet.

Thanks,
David

mlchung commented:
Hi Aleksei,

I'll review it next week as I'm taking some time off this week.

Mandy

voitylov commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -456,20 +465,109 @@ public void run() {
+            if (++counter < 0) // overflow
+                throw new Error("Maximum lock count exceeded");
+        }
+
```
Peter, the updated version checks if counter hits MAX_VALUE before incrementing. It also means the counter can't underflow, so that check is removed.

plevart replied:
This is good now.

voitylov commented:
Thanks Mandy, that would be great. Waiting on your and other core-libs members reviews!

plevart reviewed (approved):
The changes look good, Aleksei.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:
```diff
@@ -0,0 +1,159 @@
+            try {
+                this.object = fromClass.newInstance();
+                this.method = fromClass.getDeclaredMethod("loadLibrary");
+            } catch (Exception error) {
```
Nit: `s/Exception/ReflectiveOperationException/` as ROE is the specific checked exception you want to catch here.

voitylov replied:
Fixed.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:
```diff
@@ -0,0 +1,159 @@
+                this.object = fromClass.newInstance();
+                this.method = fromClass.getDeclaredMethod("loadLibrary");
+            } catch (Exception error) {
+                throw new Error(error);
```
Error is fine.  Most tests throw `RuntimeException` that can be another choice.

voitylov replied:
Yes, used RuntimeException.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:
```diff
@@ -0,0 +1,159 @@
+        public void run() {
+            try {
+                method.invoke(object);
+            } catch (Exception error) {
```
Same here to catch the `ReflectiveOperationException`.

voitylov replied:
Same here.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:
```diff
@@ -0,0 +1,159 @@
+        clazz = null;
+        threads = null;
+        exceptions.clear();
+        waitForUnload(wClass);
```
`jdk.test.lib.util.ForceGC` can be used here.  You can reference test/jdk/java/lang/ClassLoader/nativeLIbrary/NativeLibraryTest.java.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnloadTest.java:
```diff
@@ -0,0 +1,104 @@
+/*
+ * @test
+ * @bug 8266310
+ * @summary deadlock while loading the JNI code
```
Please update `@summary` with a description of the test (rather than the synposis of the bug)

voitylov replied:
Updated.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:
```diff
@@ -0,0 +1,159 @@
+/*
+ * @test
+ * @bug 8266310
+ * @summary deadlock while loading the JNI code
```
please update `@summary` with a description of the test (rather than the synposis of the bug).

voitylov replied:
Updated.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/TestLoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,237 @@
+/*
+ * @test
+ * @bug 8266310
+ * @summary deadlock while loading the JNI code
```
please update `@summary` with a description of the test (rather than the synposis of the bug).

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/TestLoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    private static OutputAnalyzer genKey() throws Throwable {
+        runCommandInTestClassPath("rm", "-f", KEYSTORE);
```
Can you use `jdk.test.lib.util.FileUtils` instead of exec `rm -f`.

voitylov replied:
Done.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/TestLoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    private static OutputAnalyzer createJar(String outputJar, String... classes) throws Throwable {
+        String jar = JDKToolFinder.getJDKTool("jar");
```
You can consider using `jdk.test.lib.util.JarUtils` to reduce the test execution time so that it can create the jar without exec'ing another process.

voitylov replied:
I tried, but this doesn't work that well unfortunately with JarUtils.createJar() as it only produces a valid jar-file capable of triggering the issue when the compiled class files (jtreg @build output) are in the same directory as the current directory of the process calling JarUtils.createJar(). Jtreg @build outputs the compiled classes to a separate directory. Creation of a new jar process allows to change the current directory so that the relative paths to the class files are such that it would form the valid output jar-file. Alternatively we could copy the class files to the current directory prior to creating the jar-file, but that would introduce an extra step. Would you mind me keeping the createJar() in the test?

mlchung replied:
Is `jar -C <dir>` option what you want?   FYI.  `jar` tool is a tool provider which is a simple way to run `jar` tool in process (lookup the tool provider by calling `ToolProvider.findFirst("jar")`).

Yes I am fine with what you have if this does not work.

mlchung reviewed (commented):
The change looks good in general with a minor suggestion to rename `NativeLibraryContext::get` to `NativeLibraryContext::current` which would be clearer as it returns the current native library context (of the current thread).

voitylov commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -202,14 +204,13 @@ private NativeLibrary loadLibrary(Class<?> fromClass, String name, boolean isBui
              * We use a static stack to hold the list of libraries we are
-             * loading because this can happen only when called by the
-             * same thread because this block is synchronous.
+             * loading, so that each thread maintains its own stack.
```
Done.

voitylov commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -202,14 +204,13 @@ private NativeLibrary loadLibrary(Class<?> fromClass, String name, boolean isBui
+             * a different class loader, we raise UnsatisfiedLinkError.
              */
-            for (NativeLibraryImpl lib : nativeLibraryContext) {
+            for (NativeLibraryImpl lib : NativeLibraryContext.get()) {
```
Done.

voitylov commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/LoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,72 @@
+                } catch (ClassNotFoundException |
+                         InstantiationException |
+                         IllegalAccessException ignore) {
+                    System.out.println("Class Class1 not found.");
```
Fixed.

voitylov commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/LoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,72 @@
+                    Class c2 = Class.forName("p.Class2");
+                    System.out.println("Signed jar loaded.");
+                } catch (ClassNotFoundException ignore) {
+                    System.out.println("Class Class2 not found.");
```
Fixed.

voitylov commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/TestLoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,237 @@
+        });
+        try {
+            return future.get(1000, TimeUnit.MILLISECONDS);
+        } catch (Exception ignoreAll) {
```
If this threw an exception, it wouldn't be possible to collect all available data from the process which is not yet completed. Throwing an exception would defeat the purpose of readAvailable() method.

voitylov commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/TestLoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,237 @@
+                .shouldHaveExitValue(0);
+
+        runCommandInTestClassPath("rm", "-f", "*.jar")
+                .shouldHaveExitValue(0);
```
Fixed.

voitylov commented:
Mandy, thank you for review! I used all of your suggestions except for two places. Please see a responses in the conversation.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/LoadLibraryDeadlock.java:
```diff
@@ -0,0 +1,74 @@
+            public void run() {
+                try {
+                    // an instance of unsigned class that loads a native library
+                    Class c1 = Class.forName("Class1");
```
nit: `s/Class/Class<?>/ ` avoid raw type (same in line 58)

voitylov replied:
Changed as suggested.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:
```diff
@@ -0,0 +1,160 @@
+        @Override
+        public Class<?> loadClass(String name) throws ClassNotFoundException {
+            synchronized (getClassLoadingLock(name)) {
+                Class clazz = findLoadedClass(name);
```
nit: `s/Class/Class<?>/`

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryUnload/p/Class1.java:
```diff
@@ -0,0 +1,43 @@
+        try {
+            System.loadLibrary("loadLibraryUnload");
+            System.out.println("Native library loaded from Class1.");
+        } catch (Exception ignore) {
```
should this exception just be thrown?

mlchung reviewed (commented):
Thanks for making the change.  I have a few minor comments and looking good.

I think the synopsis of the JBS issue should be updated to reflect this problem.  What about "deadlock between System::loadLibrary and JNI FindClass loading another class"?

voitylov commented:
Thanks Mandy. I used your suggestions in the updated version.

mlchung commented on test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/TestLoadLibraryDeadlock.java:
```diff
@@ -84,12 +88,15 @@ private static OutputAnalyzer genKey() throws Throwable {
+        List<String> args = new ArrayList<>();
+        Collections.addAll(args, "cvf", Paths.get(testClassPath, outputJar).toString());
+        for (String c : classes) {
+            Collections.addAll(args, "-C", testClassPath, c);
```
I believe only one single `-C` option is good enough for this case since the classes are all in one single directory.

voitylov replied:
Contrary to the man which says "-C DIR Changes the specified directory and includes the files specified at the end of the command line.", here is how it works:

```
bash-3.2$ find .
.
./dir
./dir/B.class
./dir/A.class
bash-3.2$ jar cvf a.jar -C dir A.class B.class
B.class : no such file or directory
added manifest
adding: A.class(in = 176) (out= 149)(deflated 15%)
```

and no jar is produced. Lines 614 and 647 of src/jdk.jartool/share/classes/sun/tools/jar/Main.java explain why. IIRC the intent was to mimic tar -C behaviour where -C applies to all files that follow, but jar does not work that way currently. 

It's probably worth reaching consensus that it's a problem worth fixing first (a scary one to fix by the way, provided how many tools could rely on the current behaviour). Alternatively, man could be adjusted to reflect the current order. I could work on that as a follow up.

mlchung replied:
You are probably right.  I have no more comment.

mlchung commented:
Looks good.  Do you want to fix this in JDK 17?  You will need to create a new PR for jdk17 repo and withdraw this one.

voitylov commented:
Thanks Mandy for the thorough review, I'll definitely do that. Shall we wait for Alan opinion as well?

mlchung commented:
> Thanks Mandy for the thorough review, I'll definitely do that. Shall we wait for Alan opinion as well?

I think you can proceed posting a PR for jdk17 as the new version has addressed Alan's concern of using ThreadLocal.  Alan can review the new PR.

voitylov commented:
/integrate

voitylov commented:
/integrate

AlexanderScherbatiy commented:
/sponsor
